### PR TITLE
[IOTDB-5106] Parse PathPattern to Deterministic Finite Automate

### PR DESCRIPTION
**Pull request by Cpaulyz**
## Description

This PR implements the parse of PathPattern to DFA and is used in the `AbstractTreeVisitor`.

Compared to NFA, using DFA gives better performance in the search process, but requires additional construction overhead.The FA generation rules are as follows: DFA is used when and only when a pathpattern exists** and there is no regex expression, otherwise NFA is used by default. In addition, to save construction overhead, a cache with a default size of 20 is used in the `FAFActory`.

Details of the design can be found at: https://apache-iotdb.feishu.cn/docx/DjrAdDM3woXcn1x5Phlcyveznqe

The performance test results given are as follows:

For complex tree search:
  | CalculateTime(ms) | InitialTime(ms) | SumTime(ms)
-- | -- | -- | --
DFA | 577 | 3 | 581
NFA | 875 | 0 | 875

For simple tree search:
  | CalculateTime(ms) | InitialTime(ms) | SumTime(ms)
-- | -- | -- | --
DFA(cached) | 148 | 246 | 395
DFA(no-cached) | 163 | 463 | 626
NFA | 192 | 241 | 434
Hybrid | 155 | 256 | 412

### CODE REVIEW

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/transition/DFARegexTransition.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.commons.path.fa.dfa.transition;
+
+import java.util.regex.Pattern;
+
+public class DFARegexTransition extends AbstractDFATransition {
```
Comment (MarcosZyk) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/transition/DFARegexTransition.java:23]: Is this class Deprecated? When encountering regex scenarios, it is SimpleNFA that will be used.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/DFAGraph.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    Closure curClosure = new Closure(closureSize);
+    curClosure.addState(curState);
+    closureStateMap.put(curClosure, curState);
+    Stack<Closure> closureStack = new Stack<>();
+    closureStack.push(curClosure);
+
+    // construct DFA
+    while (!closureStack.isEmpty()) {
+      curClosure = closureStack.pop();
+      for (IFATransition transition : transitions) {
```
Comment (MarcosZyk) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/DFAGraph.java:59]: Add annotation ```subset construction``` referred to fundamentals of compiling.

File: node-commons/src/test/java/org/apache/iotdb/commons/path/PatternDFATest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      Assert.assertFalse(checkMatchUsingDFA(new PartialPath(pattern), path));
+    }
+  }
+
+  // TODO: remove this
+  @Test
+  @Ignore
+  public void test() throws IllegalPathException {
+
+    Assert.assertFalse(
```
Comment (MarcosZyk) [inline node-commons/src/test/java/org/apache/iotdb/commons/path/PatternDFATest.java:92]: Remove this.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/NFAGraph.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  private final List<IFAState> nfaStateList = new ArrayList<>();
+
+  // [transitionIndex][stateIndex]List<IFAState>
+  private final List<IFAState>[][] nfaTransitionTable;
+
+  public NFAGraph(
+      PartialPath pathPattern, boolean isPrefix, Map<String, IFATransition> transitionMap) {
+    nfaTransitionTable = new List[transitionMap.size()][pathPattern.getNodeLength() + 1];
+    // init start state, curNodeIndex=0
+    int curStateIndex = 0;
```
Comment (MarcosZyk) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/NFAGraph.java:45]: Give an example of what will the NFAGraph be like when input pattern xxxxxx.

File: node-commons/src/test/java/org/apache/iotdb/commons/path/PatternDFATest.java (hunk 3 above)
Comment (Cpaulyz) [reply node-commons/src/test/java/org/apache/iotdb/commons/path/PatternDFATest.java]: done

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/DFAGraph.java (hunk 2 above)
Comment (Cpaulyz) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/DFAGraph.java]: done

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/transition/DFARegexTransition.java (hunk 1 above)
Comment (Cpaulyz) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/transition/DFARegexTransition.java]: removed

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/NFAGraph.java (hunk 4 above)
Comment (Cpaulyz) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/dfa/graph/NFAGraph.java:49]: add Demo in PatternDFATest#printFASketch

**Review by MarcosZyk — APPROVED** [review]
Fantastic work!

**Pull request merged**